contrib/endpoints/src/api_manager/config_manager_impl.cc

```diff
@@ -0,0 +1,313 @@
+        global_context_->env()->LogError(msg);
+        config_roollout_callback_(utils::Status(Code::ABORTED, msg),
+                                  empty_configs);
+        return;
+      }
+
+      // TODO(jaebong) config_id should not be empty for the first version
+      // This part will be removed after the rollouts feature added
+      if (global_context_->config_id().empty()) {
+        std::string msg = "API config_id not specified in configuration files";
+        global_context_->env()->LogError(msg);
+        config_roollout_callback_(utils::Status(Code::ABORTED, msg),
+                                  empty_configs);
+        return;
+      }
+
+      // Check config_id is invalid
+      if (!isValidConfigId(global_context_->config_id())) {
+        std::string msg = "Invalid config_id: " + global_context_->config_id();
+        global_context_->env()->LogError(msg);
+        config_roollout_callback_(utils::Status(Code::ABORTED, msg),
+                                  empty_configs);
+        return;
+      }
+
+      // Fetch service account token
+      GlobalFetchServiceAccountToken(global_context_,
+                                     on_fetch_auth_token_callback_);
+    } else if (status.code() == Code::INTERNAL) {
+      // Failed to fetch meta data
+      config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                                empty_configs);
+    } else {
+      // Still downloading metata data. Try to fech again
+      GlobalFetchGceMetadata(global_context_, on_fetch_metadata_callback_);
+    }
+  };
+}
+
+std::function<void(utils::Status)>
+ConfigManagerImpl::get_fetch_auth_token_callback() {
+  return [this](const utils::Status& status) {
+    if (status.ok()) {
+      if (global_context_->service_account_token()) {
+        // register auth token for servicemanagement services
+        global_context_->service_account_token()->SetAudience(
+            auth::ServiceAccountToken::JWT_TOKEN_FOR_SERVICEMANAGEMENT_SERVICES,
+            service_management_url_ + kServiceManagementServiceManager);
+      }
+
+      // Fetch configs from the Inceptions
+      std::vector<std::pair<std::string, int>> rollouts = {
+          {global_context_->config_id(), 0}};
+      std::vector<std::pair<std::string, int>> configs;
+      fetch_configs(rollouts, 0, configs);
+    } else if (status.code() == Code::INTERNAL) {
+      // Failed to fetch auth token.
+      config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                                empty_configs);
+    } else {
+      // Still downloading auth token.  Retry.
+      GlobalFetchServiceAccountToken(global_context_,
+                                     on_fetch_auth_token_callback_);
+    }
+  };
+}
+
+void ConfigManagerImpl::fetch_configs(
+    std::vector<std::pair<std::string, int>> rollouts, std::size_t index,
+    std::vector<std::pair<std::string, int>> configs) {
```

Comment:
It is better to define a structure and pass that data around by using shared_ptr for fetching configs.  not passing two vectors around in the function arguments.